File: modulo-java/src/com/fabrica/muebles/dao/ProduccionDAO.java
package com.fabrica.muebles.dao;

import com.fabrica.muebles.modelo.Produccion;
import com.fabrica.muebles.util.ConexionBD;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

/**
 * DAO básico para operaciones CRUD de Producción
 * Autor: William Alonso Samaca Lopez
 */
public class ProduccionDAO {

    // Insertar
    public boolean insertarProduccion(Produccion produccion) {
        String sql = "INSERT INTO produccion (nombre_producto, cantidad, fecha_inicio, fecha_fin, estado) VALUES (?, ?, ?, ?, ?)";
        try (Connection con = ConexionBD.getConexion();
             PreparedStatement ps = con.prepareStatement(sql)) {

            ps.setString(1, produccion.getNombreProducto());
            ps.setInt(2, produccion.getCantidad());
            ps.setDate(3, produccion.getFechaInicio());
            ps.setDate(4, produccion.getFechaFin());
            ps.setString(5, produccion.getEstado());
            return ps.executeUpdate() > 0;

        } catch (SQLException e) {
            System.err.println("Error al insertar producción: " + e.getMessage());
            return false;
        }
    }

    // Consultar todos
    public List<Produccion> consultarTodos() {
        List<Produccion> lista = new ArrayList<>();
        String sql = "SELECT * FROM produccion ORDER BY id DESC";

        try (Connection con = ConexionBD.getConexion();
             Statement st = con.createStatement();
             ResultSet rs = st.executeQuery(sql)) {

            while (rs.next()) {
                Produccion p = new Produccion(
                    rs.getInt("id"),
                    rs.getString("nombre_producto"),
                    rs.getInt("cantidad"),
                    rs.getDate("fecha_inicio"),
                    rs.getDate("fecha_fin"),
                    rs.getString("estado")
                );
                lista.add(p);
            }

        } catch (SQLException e) {
            System.err.println("Error al consultar producciones: " + e.getMessage());
        }
        return lista;
    }

    // Consultar por ID
    public Produccion consultarPorId(int id) {
        String sql = "SELECT * FROM produccion WHERE id = ?";
        try (Connection con = ConexionBD.getConexion();
             PreparedStatement ps = con.prepareStatement(sql)) {

            ps.setInt(1, id);
            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) {
                    return new Produccion(
                        rs.getInt("id"),
                        rs.getString("nombre_producto"),
                        rs.getInt("cantidad"),
                        rs.getDate("fecha_inicio"),
                        rs.getDate("fecha_fin"),
                        rs.getString("estado")
                    );
                }
            }
        } catch (SQLException e) {
            System.err.println("Error al consultar producción: " + e.getMessage());
        }
        return null;
    }

    // Actualizar
    public boolean actualizarProduccion(Produccion produccion) {
        String sql = "UPDATE produccion SET nombre_producto=?, cantidad=?, fecha_inicio=?, fecha_fin=?, estado=? WHERE id=?";
        try (Connection con = ConexionBD.getConexion();
             PreparedStatement ps = con.prepareStatement(sql)) {

            ps.setString(1, produccion.getNombreProducto());
            ps.setInt(2, produccion.getCantidad());
            ps.setDate(3, produccion.getFechaInicio());
            ps.setDate(4, produccion.getFechaFin());
            ps.setString(5, produccion.getEstado());
            ps.setInt(6, produccion.getId());
            return ps.executeUpdate() > 0;

        } catch (SQLException e) {
            System.err.println("Error al actualizar producción: " + e.getMessage());
            return false;
        }
    }

    // Finalizar producción
    public boolean finalizarProduccion(int id) {
        String sql = "UPDATE produccion SET estado='Finalizado', fecha_fin=CURDATE() WHERE id=?";
        try (Connection con = ConexionBD.getConexion();
             PreparedStatement ps = con.prepareStatement(sql)) {

            ps.setInt(1, id);
            return ps.executeUpdate() > 0;

        } catch (SQLException e) {
            System.err.println("Error al finalizar producción: " + e.getMessage());
            return false;
        }
    }

    // Eliminar
    public boolean eliminarProduccion(int id) {
        String sql = "DELETE FROM produccion WHERE id=?";
        try (Connection con = ConexionBD.getConexion();
             PreparedStatement ps = con.prepareStatement(sql)) {

            ps.setInt(1, id);
            return ps.executeUpdate() > 0;

        } catch (SQLException e) {
            System.err.println("Error al eliminar producción: " + e.getMessage());
            return false;
        }
    }
}

File: src/com/fabrica/muebles/dao/produccion_dao.py
from com.fabrica.muebles.modelo.produccion import Produccion
from com.fabrica.muebles.util.conexion_bd import ConexionBD


class ProduccionDAO:
    """DAO para operaciones CRUD de Producción"""

    def insertar_produccion(self, produccion):
        sql = """INSERT INTO produccion 
                 (codigo_mueble, nombre_cliente, tipo_documento_cliente, numero_documento_cliente, 
                  nombre_producto, cantidad, fecha_inicio, fecha_fin, estado, observaciones, ruta_archivo) 
                 VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)"""
        try:
            conexion = ConexionBD.get_conexion()
            cursor = conexion.cursor()
            cursor.execute(sql, (
                produccion.codigo_mueble,
                produccion.nombre_cliente,
                produccion.tipo_documento_cliente,
                produccion.numero_documento_cliente,
                produccion.nombre_producto,
                produccion.cantidad,
                produccion.fecha_inicio,
                produccion.fecha_fin,
                produccion.estado,
                produccion.observaciones,
                produccion.ruta_archivo
            ))
            conexion.commit()
            return cursor.rowcount > 0
        except Exception as e:
            print(f"Error: {e}")
            if 'conexion' in locals():
                conexion.rollback()
            return False
        finally:
            if 'cursor' in locals():
                cursor.close()
            if 'conexion' in locals():
                conexion.close()

    def consultar_todos(self):
        sql = """SELECT id, codigo_mueble, nombre_cliente, tipo_documento_cliente, numero_documento_cliente,
                        nombre_producto, cantidad, fecha_inicio, fecha_fin, estado, observaciones, ruta_archivo 
                 FROM produccion 
                 ORDER BY id DESC"""
        try:
            conexion = ConexionBD.get_conexion()
            cursor = conexion.cursor()
            cursor.execute(sql)
            return [Produccion(*fila) for fila in cursor.fetchall()]
        except Exception as e:
            print(f"Error: {e}")
            return []
        finally:
            cursor.close()
            conexion.close()

    def consultar_por_id(self, id):
        sql = """SELECT id, codigo_mueble, nombre_cliente, tipo_documento_cliente, numero_documento_cliente,
                        nombre_producto, cantidad, fecha_inicio, fecha_fin, estado, observaciones, ruta_archivo 
                 FROM produccion 
                 WHERE id = %s"""
        try:
            conexion = ConexionBD.get_conexion()
            cursor = conexion.cursor()
            cursor.execute(sql, (id,))
            fila = cursor.fetchone()
            return Produccion(*fila) if fila else None
        except Exception as e:
            print(f"Error: {e}")
            return None
        finally:
            cursor.close()
            conexion.close()

    def actualizar_produccion(self, produccion):
        sql = """UPDATE produccion 
                 SET codigo_mueble=%s, nombre_cliente=%s, tipo_documento_cliente=%s, numero_documento_cliente=%s,
                     nombre_producto=%s, cantidad=%s, fecha_inicio=%s, fecha_fin=%s, estado=%s, observaciones=%s, ruta_archivo=%s 
                 WHERE id=%s"""
        try:
            conexion = ConexionBD.get_conexion()
            cursor = conexion.cursor()
            cursor.execute(sql, (
                produccion.codigo_mueble,
                produccion.nombre_cliente,
                produccion.tipo_documento_cliente,
                produccion.numero_documento_cliente,
                produccion.nombre_producto,
                produccion.cantidad,
                produccion.fecha_inicio,
                produccion.fecha_fin,
                produccion.estado,
                produccion.observaciones,
                produccion.ruta_archivo,
                produccion.id
            ))
            conexion.commit()
            return cursor.rowcount > 0
        except Exception as e:
            print(f"Error: {e}")
            if 'conexion' in locals():
                conexion.rollback()
            return False
        finally:
            cursor.close()
            conexion.close()

    def finalizar_produccion(self, id):
        sql = "UPDATE produccion SET estado='Finalizado', fecha_fin=CURDATE() WHERE id=%s"
        try:
            conexion = ConexionBD.get_conexion()
            cursor = conexion.cursor()
            cursor.execute(sql, (id,))
            conexion.commit()
            return cursor.rowcount > 0
        except Exception as e:
            print(f"Error: {e}")
            if 'conexion' in locals():
                conexion.rollback()
            return False
        finally:
            cursor.close()
            conexion.close()

    def eliminar_produccion(self, id):
        try:
            conexion = ConexionBD.get_conexion()
            cursor = conexion.cursor()
            cursor.execute("DELETE FROM produccion WHERE id=%s", (id,))
            conexion.commit()
            return cursor.rowcount > 0
        except Exception as e:
            print(f"Error: {e}")
            if 'conexion' in locals():
                conexion.rollback()
            return False
        finally:
            cursor.close()
            conexion.close()